src/auth/bearer/UserRequest.cc

```diff
@@ -0,0 +1,264 @@
+
+    if (sz<=0) {
+        debugs(9, DBG_CRITICAL, "ERROR: Bearer Authentication Failure. Can not build helper validation request.");
+        handler(data);
+    } else if (static_cast<size_t>(sz) >= sizeof(buf)) {
+        debugs(9, DBG_CRITICAL, "ERROR: Bearer Authentication Failure. Helper request line exceeds " << sizeof(buf) << " bytes.");
+        handler(data);
+    } else {
+        user()->credentials(Auth::Pending);
+        debugs(29, 3, "token " << bearer_auth->token->b68encoded << " lookup started");
+        helperSubmit(bearerauthenticators, buf, Auth::Bearer::UserRequest::HandleReply,
+                     new Auth::StateData(this, handler, data));
+    }
+}
+
+void
+Auth::Bearer::UserRequest::authenticate(HttpRequest * aRequest, ConnStateData * conn, Http::HdrType type)
+{
+    assert(user());
+
+    // if the password is not ok, do an identity
+    if (!user() || user()->credentials() != Auth::Ok)
+        return;
+
+    // are we about to recheck the credentials externally?
+    if (user()->ttl() <= 0) {
+        debugs(29, 4, "credentials expired - rechecking");
+        user()->credentials(Auth::Unchecked);
+        return;
+    }
+
+    // we have been through the external helper, and the credentials haven't expired
+    debugs(29, 9, "user " << user()->username() << " authenticated");
+}
+
+void
+Auth::Bearer::UserRequest::HandleReply(void *data, const Helper::Reply &reply)
+{
+    Auth::StateData *r = static_cast<Auth::StateData *>(data);
+
+    debugs(29, 8, reply.reservationId << " got reply=" << reply);
+
+    if (!cbdataReferenceValid(r->data)) {
+        debugs(29, DBG_IMPORTANT, "ERROR: Bearer Authentication invalid callback data.");
+        delete r;
+        return;
+    }
+
+    Auth::UserRequest::Pointer auth_user_request = r->auth_user_request;
+    assert(auth_user_request);
+
+    // add new helper kv-pair notes to the credentials object
+    // so that any transaction using those credentials can access them
+    auth_user_request->user()->notes.appendNewOnly(&reply.notes);
+
+    assert(auth_user_request->user());
+    assert(auth_user_request->user()->auth_type == Auth::AUTH_BEARER);
+
+    switch (reply.result) {
+    case Helper::Okay: {
+        const char *userNote = reply.notes.findFirst("user");
+        const char *ttlNote = reply.notes.findFirst("ttl");
+        if (!userNote) {
+            /* protocol error */
+            fatalf("Auth::Bearer::HandleReply: *** Unsupported helper response ***, '%s'\n", reply.other().content());
+            break;
+        }
+
+        Auth::Bearer::User *usr = dynamic_cast<Auth::Bearer::User *>(auth_user_request->user().getRaw());
+        /* we're finished, user is authenticated */
+        usr->username(userNote);
+        auth_user_request->denyMessage("Login successful");
+        const int64_t ttl = (ttlNote ? strtoll(ttlNote, nullptr, 10) : -1);
+        usr->token->expires = current_time.tv_sec + ttl;
```

Comment:
We do. "ttl=-n" and the -1 default both create an entry protecting against retry attacks using that token for future traffic by this username.

Reply:
```C++
/// this token should be ignored when received passed this time
time_t expires = 0;
```

> "ttl=-n" and the -1 default both create an entry protecting against retry attacks using that token for future traffic by this username.

If the intent is to mark the token as expired, should we reset `token->expires` to zero when receiving a negative TTL? Feels wrong that we are happily setting `token->expires` to some _specific_ moment in the past (e.g., "10 seconds ago") if what we are actually trying to do is to expire that token (so that it becomes ignored, per `expires` field description). This becomes especially suspicious when we start comparing the adjusted `expires` value with other very specific moments like `usr->expiretime` on the next line.

Reply:
Sorry should have responded earlier to this. What we are supporting here with -n TTL values is that the helper can specify not just that the credentials are stale, but by how much. Which opens some useful possibilities when tuning garbage collection and credentials grace period updates in squid.conf.